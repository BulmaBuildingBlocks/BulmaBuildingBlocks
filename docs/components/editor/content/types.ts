export enum EditableContentTypes {
  Text = 'text',
  Title = 'title'
}

export interface EditableContentProps {
  /**
   * Element wrapping the Editable Content.
   */
  tag?: string;

  /**
   * Provides different editing options
   */
  type?: EditableContentTypes;

  /**
   * Class placed on the parent element.
   */
  class?: string;

  /**
   * Value that's edited.
   */
  value: string;

  /**
   * Content Editable.
   */
  editable?: boolean;
}
