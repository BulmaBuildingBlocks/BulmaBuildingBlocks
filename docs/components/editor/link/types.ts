import { Status } from '~/shared/config';

export enum LinkType {
  Link = 'Link',
  Button = 'button',
  Rounded = 'button is-rounded',
  Outlined = 'button is-outlined',
  'Rounded Outlined' = 'button is-rounded is-outlined'
}

export class EditableLink {
  constructor(props: EditableLinkProps) {
    this.label = props.label;
    this.status = props.status || Status.None;
    this.linkType = props.linkType;
    this.href = props.href;
    this.popupPlacement = props.popupPlacement;
  }

  /*
   * Label on the button.
   */
  label: string;

  /**
   * Assigned hyperlink.
   */
  href?: string;

  /**
   * Link styles.
   */
  linkType: LinkType;

  /**
   * Link styles.
   */
  status: Status;

  /**
   * Popup Placement.
   */
  popupPlacement?: string;
}

export interface EditableLinkProps {
  /**
   * Label on the button.
   */
  label: string;

  /**
   * Assigned hyperlink.
   */
  href?: string;

  /**
   * Link styles.
   */
  linkType: LinkType;

  /**
   * Link styles.
   */
  status?: Status;

  /**
   * Popup Placement.
   */
  popupPlacement?: string;
}
