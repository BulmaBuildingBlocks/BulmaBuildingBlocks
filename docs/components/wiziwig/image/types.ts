export interface EditableImageProps {
  /**
   * Key of a permalink referencing the source of the picture.
   */
  src: string;

  /**
   * Caption on the picture, used also as alternative text.
   */
  alt: string;

  /**
   * Hyperlink attached to the picture.
   */
  href?: string;

  /**
   * Picture width.
   */
  width?: number;

  /**
   * Picture height.
   */
  height?: number;

  /**
   * Popup Placement.
   */
  popupPlacement?: string;
}
