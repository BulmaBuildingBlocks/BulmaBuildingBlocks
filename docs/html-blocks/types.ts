import { VueConstructor } from 'vue';

export enum ContentTypes {
  Headers = 'headers',
  Features = 'features',
  CallToActions = 'calltoactions',
  Contents = 'contents',
  Footers = 'footers'
}

export interface Block {
  title: string;
  type?: ContentTypes;
  componentName: string;
  component: VueConstructor;
  content?: any;
  color?: string;
  container?: boolean;
}

export interface BlockListObject {
  [x: string]: VueConstructor;
}
