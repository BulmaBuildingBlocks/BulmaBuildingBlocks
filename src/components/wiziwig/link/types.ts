export interface EditableLinkProps {
  /**
   * Label on the button.
   */
  label: string;

  /**
   * Assigned hyperlink.
   */
  href?: string;

  /**
   * Link styles.
   */
  styles?: string;

  /**
   * Popup Placement.
   */
  popupPlacement?: string;
}
