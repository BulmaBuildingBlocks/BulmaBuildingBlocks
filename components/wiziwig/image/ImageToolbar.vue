<template>
  <div class="image-toolbar card">
    <header class="card-header">
      <p class="card-header-title">
        Image Configuration
      </p>
      <a class="card-header-icon" aria-label="close" @click="closeModal">
        <b-icon pack="fa" icon="times" />
      </a>
    </header>
    <div class="card-content">
      <b-field label="Image Url" placeholder="http://">
        <b-input v-model="newValue.src" />
      </b-field>
      <b-field label="Description">
        <b-input v-model="newValue.alt" />
      </b-field>
      <b-field label="Image Link">
        <b-input v-model="newValue.href" />
      </b-field>
    </div>
  </div>
</template>

<script lang="ts">
import {
  Vue,
  Component,
  PropSync
} from '~/node_modules/nuxt-property-decorator';
import { EditableImageProps } from '~/components/wiziwig/image/types';

/***
 * Editting Features:
 * Upload and replace image
 * Alt tag
 * Link
 */

@Component
export default class ImageToolbar extends Vue {
  @PropSync('value', { type: Object }) newValue!: EditableImageProps;

  closeModal() {
    this.$emit('closeModal');
  }
}
</script>
