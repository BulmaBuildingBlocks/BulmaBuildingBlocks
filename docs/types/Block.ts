import { VueConstructor } from 'vue';

export interface Block {
  title: string;
  component: VueConstructor;
  content?: any;
  color?: string;
  container?: boolean;
}

export interface BlockListObject {
  [x: string]: VueConstructor;
}
