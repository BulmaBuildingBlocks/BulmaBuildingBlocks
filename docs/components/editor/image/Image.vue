<script lang="ts">
import { CreateElement } from 'vue';
import { Vue, Component, Prop, PropSync } from 'nuxt-property-decorator';
import ImageToolbar from './ImageToolbar.vue';
import PopupModal from '~/components/PopupModal.vue';

import { EditableImageProps } from '~/components/editor/image/types';
import ClickOutside from '~/directives/click-outside';

/***
 * Editting Features:
 * Upload and replace image
 * Alt tag
 * Link
 */

@Component({
  directives: {
    ClickOutside
  }
})
export default class EditableImage extends Vue {
  @Prop(Boolean) editable!: boolean;
  @PropSync('value', { type: Object }) newValue!: EditableImageProps;

  modalOpen = false;

  toggleModal(event: Event) {
    event.preventDefault();

    this.modalOpen = !this.modalOpen;
  }

  closeModal() {
    this.modalOpen = false;
  }

  render(createElement: CreateElement) {
    const image = createElement('img', {
      attrs: {
        src: this.newValue.src,
        alt: this.newValue.alt,
        width: this.newValue.width,
        height: this.newValue.height
      },
      on: {
        click: this.toggleModal
      }
    });

    const imageWithLink = createElement(
      'a',
      {
        attrs: {
          href: this.newValue.href
        }
      },
      [image]
    );

    const imageToolbar = createElement(ImageToolbar, {
      props: {
        value: this.newValue
      },
      on: {
        closeModal: () => {
          this.closeModal();
        },
        'update:value': (value: EditableImageProps) => {
          this.newValue = value;
        }
      }
    });

    const popupModal = createElement(PopupModal, {
      scopedSlots: {
        reference: () => {
          return this.newValue.href ? imageWithLink : image;
        },
        default: () => {
          return imageToolbar;
        }
      },
      directives: [
        {
          name: 'click-outside',
          value: { handler: this.closeModal }
        }
      ],
      props: {
        show: this.editable && this.modalOpen,
        options: {
          placement: this.newValue.popupPlacement || 'right'
        },
        modifiers: [
          {
            name: 'flip',
            options: {
              fallbackPlacements: ['left', 'top', 'right', 'bottom']
            }
          }
        ]
      }
    });

    return popupModal;
  }
}
</script>
