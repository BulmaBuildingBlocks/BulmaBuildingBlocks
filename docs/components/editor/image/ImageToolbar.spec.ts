import { shallowMount, Wrapper } from '@vue/test-utils';
import ImageToolbar from '~/components/editor/image/ImageToolbar.vue';
import { EditableImageProps } from '~/components/editor/image/types';

describe('ImageToolbar.vue', () => {
  let wrapper: Wrapper<ImageToolbar & { [key: string]: any }>;

  beforeEach(() => {
    // Mount the component
    const props: EditableImageProps = {
      alt: '',
      src: ''
    };

    wrapper = shallowMount(ImageToolbar, {
      propsData: { value: props }
    });
  });

  afterEach(() => {
    wrapper.destroy();
  });

  it('has the expected html structure', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  test('is a Vue instance', () => {
    expect(wrapper.isVueInstance).toBeTruthy();
  });

  // Inspect the raw component options
  test('has a close modal function', () => {
    expect(typeof wrapper.vm.closeModal).toBe('function');
  });

  // Emit event from close modal
  test('has a close modal function', async () => {
    // call function to emit event
    wrapper.vm.closeModal();

    await wrapper.vm.$nextTick(); // Wait until $emits have been handled

    expect(wrapper.emitted().closeModal).toBeTruthy();
  });
});
