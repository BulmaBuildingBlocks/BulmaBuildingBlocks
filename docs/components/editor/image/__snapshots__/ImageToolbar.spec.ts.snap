// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageToolbar.vue has the expected html structure 1`] = `
<div
  class="image-toolbar card"
>
  <header
    class="card-header"
  >
    <p
      class="card-header-title"
    >
      
      Image Configuration
    
    </p>
     
    <a
      aria-label="close"
      class="card-header-icon"
    >
      <b-icon-stub
        icon="times"
        pack="fa"
      />
    </a>
  </header>
   
  <div
    class="card-content"
  >
    <b-field-stub
      addons="true"
      label="Image Url"
      placeholder="http://"
    >
      <b-input-stub
        customclass=""
        hascounter="true"
        type="text"
        usehtml5validation="true"
        value=""
      />
    </b-field-stub>
     
    <b-field-stub
      addons="true"
      label="Description"
    >
      <b-input-stub
        customclass=""
        hascounter="true"
        type="text"
        usehtml5validation="true"
        value=""
      />
    </b-field-stub>
     
    <b-field-stub
      addons="true"
      label="Image Link"
    >
      <b-input-stub
        customclass=""
        hascounter="true"
        type="text"
        usehtml5validation="true"
      />
    </b-field-stub>
  </div>
</div>
`;
